File: python/find_mode_in_binary_search_tree.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.count = 1
        self.previous = None
        self.max = 0

    def find_mode(self, root: TreeNode, modes: List[int]):
        if root is None:
            return
        self.find_mode(root.left, modes)
        if self.previous:
            self.count = self.count + 1 if root.val == self.previous.val else 1
        if self.count > self.max:
            self.max = self.count
            modes.clear()
            modes.append(root.val)
        elif self.count == self.max:
            modes.append(root.val)
        self.previous = root
        self.find_mode(root.right, modes)

    def findMode(self, root: TreeNode) -> List[int]:
        if root is None: return []
        result = []
        self.find_mode(root, result)
        return result

File: src/FindModeInBinarySearchTree.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FindModeInBinarySearchTree {
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private static TreeNode previous;
    private static int count = 1;
    private static int mode = 0;

    public static int[] findMode(TreeNode root) {
        if (root == null) {
            return new int[0];
        }

        List<Integer> modes = new ArrayList<>();
        findMode(root, modes);
        int[] result = new int[modes.size()];
        for(int index = 0 ; index < modes.size() ; index++) {
            result[index] = modes.get(index);
        }
        previous = null;
        count = 1;
        mode = 0;
        return result;
    }

    private static void findMode(TreeNode root, List<Integer> modes) {
        if (root == null) {
            return;
        }
        findMode(root.left, modes);
        if (previous != null) {
            count = root.val == previous.val ? count + 1 : 1;
        }
        if (count > mode) {
            mode = count;
            modes.clear();
            modes.add(root.val);
        } else if (count == mode) {
            modes.add(root.val);
        }
        previous = root;
        findMode(root.right, modes);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(0);
        System.out.println(Arrays.toString(findMode(root)));
    }
}
